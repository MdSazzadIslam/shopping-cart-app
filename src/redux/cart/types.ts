export enum CartActionTypes {
  ADD_TO_CART = 'ADD_TO_CART',
  ADD_TO_CART_FAILURE = 'ADD_TO_CART_FAILURE',
  REMOVE_FROM_CART = 'REMOVE_FROM_CART',
  REMOVE_FROM_CART_FAILURE = 'REMOVE_FROM_CART_FAILURE',
  GET_CART_REQUEST = 'GET_CART_REQUEST',
  GET_CART_SUCCESS = 'GET_CART_SUCCESS',
  GET_CART_ERROR = 'GET_CART_ERROR',
}
