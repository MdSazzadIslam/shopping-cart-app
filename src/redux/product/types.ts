export enum ProductActionTypes {
  GET_PRODUCT_REQUEST = "GET_PRODUCT_REQUEST",
  GET_PRODUCT_SUCCESS = "GET_PRODUCT_SUCCESS",
  GET_PRODUCT_ERROR = "GET_PRODUCT_ERROR",
}
