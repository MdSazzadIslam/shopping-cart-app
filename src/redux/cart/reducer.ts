import { Reducer } from 'redux';

import { CartActionTypes } from './types';
import { CartState } from '../../types/cart';

export const initialState: CartState = {
  data: {
    userId: Date.now().toString(36) + Math.random().toString(36).substring(2),
    items: [],
  },
  errors: undefined,
  loading: false,
};

const reducer: Reducer<CartState> = (state = initialState, action) => {
  switch (action.type) {
    case CartActionTypes.GET_CART_REQUEST: {
      return { ...state, loading: true };
    }
    case CartActionTypes.GET_CART_SUCCESS: {
      return { ...state, loading: false, data: action.payload };
    }
    case CartActionTypes.GET_CART_ERROR: {
      return { ...state, loading: false, errors: action.payload };
    }
    case CartActionTypes.ADD_TO_CART: {
      const data = state.data.items[0];
      const qty = 0;
      debugger;
      if (data) {
        if (
          data.id === action.payload.id &&
          data.coverage === action.payload.coverage
        )
          return {
            errors: state.errors,
            loading: state.loading,
            data: {
              ...state.data,
              items: [...state.data.items, action.payload],
            },
          };
      }
      return {
        errors: state.errors,
        loading: state.loading,
        data: {
          ...state.data,
          id: state.data.userId,
          items: [...state.data.items, action.payload],
        },
      };
    }
    case CartActionTypes.ADD_TO_CART_FAILURE: {
      return { ...state, loading: false, errors: action.payload };
    }

    case CartActionTypes.REMOVE_FROM_CART: {
      console.log(action.payload);
      return {
        errors: state.errors,
        loading: state.loading,
        data: {
          ...state.data,
          id: state.data.userId,
          items: state.data.items.filter((item) => item.id !== action.payload),
        },
      };
    }
    case CartActionTypes.REMOVE_FROM_CART_FAILURE: {
      return { ...state, loading: false, errors: action.payload };
    }

    default: {
      return state;
    }
  }
};

export { reducer as cartReducer };
