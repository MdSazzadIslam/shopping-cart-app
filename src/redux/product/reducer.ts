import { Reducer } from "redux";
import { ProductActionTypes } from "./types";
import { ProductState } from "../../types/product";

/**
 * Reducer, a part of the redux flow.
 *
 * Reducers decide how the state changes in response to an action. Reducers handle the
 * entire state of the application as a single object. The reducer takes the previous state
 * and an action to produce new state.
 *
 * Important! Reducer must never mutate state. No side effects or API calls.
 *
 *   .....................                                   ---------------
 *   .   action creator  . -->     dispatch(action)    -->   |   reducer   |
 *   .....................                                   ---------------
 *            /\                                                    |
 *            |                                                     \/
 *       ............                                         .............
 *       .   view   .     <--  state passed as props  <--     .   store   .
 *       ............                                         .............
 *
 */

/** The complete structure of the application state
 *
 * In redux, the state is never mutated directly.
 * For this reason, the ProductState state interface declares all
 * members as readonly to gain compile-time checks
 * against mutation.
 */

export const initialState: ProductState = {
  data: [],
  errors: undefined,
  loading: false,
};

/** The redux reducer for product. Handles product actions. */

const reducer: Reducer<ProductState> = (state = initialState, action) => {
  switch (action.type) {
    case ProductActionTypes.GET_PRODUCT_REQUEST: {
      return { ...state, loading: true };
    }
    case ProductActionTypes.GET_PRODUCT_SUCCESS: {
      console.log("action payload", action.payload);
      return { ...state, loading: false, data: action.payload };
    }
    case ProductActionTypes.GET_PRODUCT_ERROR: {
      return { ...state, loading: false, errors: action.payload };
    }
    default: {
      // return the previous state on any unknown action
      return state;
    }
  }
};

export { reducer as productReducer };
