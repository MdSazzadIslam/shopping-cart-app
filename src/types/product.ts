export type ProductProps = {
  id: string;
  name: string;
  coverage: string;
  risk: string;
  price: number;
  image: string;
  description: string;
};

export type getProducts = () => void;
export type AddProduct = (newProduct: string) => void;
export type DeleteProduct = (id: string) => void;

export interface ProductState {
  readonly loading: boolean;
  readonly data: ProductProps[];
  readonly errors?: string;
}
