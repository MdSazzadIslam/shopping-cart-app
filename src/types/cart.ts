import { ProductProps } from './product';
export type CartProps = {
  items: ProductProps[];
};

export interface CartState {
  readonly loading: boolean;
  readonly data: CartProps;
  readonly errors?: string;
}
