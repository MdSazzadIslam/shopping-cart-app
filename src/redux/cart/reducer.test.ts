import { CartActionTypes } from './types';
import { initialState } from './reducer';
import { cartReducer } from './reducer';

describe('The cart reducer', () => {
  it('returns the input state for an unknown action', () => {
    const state = cartReducer(initialState, { type: undefined });
    expect(state).toEqual(initialState);
  });

  /*   it('should return the initial state', () => {
    expect(cartReducer(undefined, {})).toMatchSnapshot();
  }); */

  it('should handle CART_REQUEST', () => {
    expect(
      cartReducer(initialState, {
        type: CartActionTypes.GET_CART_SUCCESS,
      })
    ).toMatchSnapshot();
  });
});
