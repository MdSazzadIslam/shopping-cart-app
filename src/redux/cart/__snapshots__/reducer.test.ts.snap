// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The cart reducer should handle CART_REQUEST 1`] = `
Object {
  "data": undefined,
  "errors": undefined,
  "loading": false,
}
`;
